jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -333,109 +333,116 @@ public ByteBuffer onUpgradeFrom()
                 // If there was a connection upgrade, the other
                 // connection took over, nothing more to do here.
                 if (getEndPoint().getConnection() != this)
+                {
+                    releaseRequestBuffer();
                     break;
+                }
 
-                // Handle channel event. This will only be true when the headers of a request have been received.
+                // The headers of a request have been received.
                 if (handle)
                 {
                     Request request = _httpChannel.getRequest();
                     if (LOG.isDebugEnabled())
                         LOG.debug("HANDLE {} {}", request, this);
 
-                    // handle the request by running the task obtained from onRequest
+                    // Handle the request by running the task.
                     _handling.set(true);
                     Runnable onRequest = _onRequest;
                     _onRequest = null;
                     onRequest.run();
 
-                    // If the _handling boolean has already been CaS'd to false, then stream is completed and we are no longer
-                    // handling, so the caller can continue to fill and parse more connections.  If it is still true, then some
-                    // thread is still handling the request and they will need to organize more filling and parsing once complete.
+                    // If the CaS succeeds, then some thread is still handling the request.
+                    // If the CaS fails, then stream is completed, we are no longer handling,
+                    // so the caller can continue to fill and parse more connections.
                     if (_handling.compareAndSet(true, false))
                     {
                         if (LOG.isDebugEnabled())
                             LOG.debug("request !complete {} {}", request, this);
+                        // Cannot release the request buffer here, because the
+                        // application may read concurrently from another thread.
+                        // The request buffer will be released by the application
+                        // reading the request content, or by the implementation
+                        // trying to consume the request content.
                         break;
                     }
 
-                    // If the request is complete, but has been upgraded, then break
+                    // If there was an upgrade, release and return.
                     if (getEndPoint().getConnection() != this)
                     {
                         if (LOG.isDebugEnabled())
                             LOG.debug("upgraded {} -> {}", this, getEndPoint().getConnection());
+                        releaseRequestBuffer();
                         break;
                     }
                 }
+                else if (filled == 0)
+                {
+                    releaseRequestBuffer();
```

Comment:
Add an assert to make it clear we know the buffer is empty here because parsing is fully consuming.
```suggestion
                    assert isRequestBufferEmpty();
                    releaseRequestBuffer();
```